### PR TITLE
Export audio output as opus - Iss 27

### PR DESCRIPTION
**Pull request by DerouineauNicolas**
This PR is aiming to implement feature described in issue #27.

The audio output of the WebAudio graph is exported through a mediaRecorder node. If the record is trigerred through the button, all audio chunks are redirected through a Blob and then are dynamically added to the page as a downloable and playable audio file (ogg).

This can only be used as a live recorder (ie: the sequencer must be in a play state).

### CODE REVIEW

File: static/javascripts/app.js [hunk 1]
```diff
@@ -206,8 +232,22 @@ function init() {
...
 }
 
+
+
 function initializeAudioNodes() {
   context = new webkitAudioContext();
+  recordingDest = context.createMediaStreamDestination();
+  mediaRecorder = new MediaRecorder(recordingDest.stream);
+
+  mediaRecorder.ondataavailable = function(evt) {
```
Comment (xavierfav) [inline static/javascripts/app.js]: I think we should avoid declaring function in the initializer function of all the nodes. It makes it hard to follow, improve or debug.

File: views/index.ejs [hunk 2]
```diff
@@ -48,6 +48,9 @@
         <button id="play-pause" type="button" class="btn btn-success">
           <span class="glyphicon glyphicon-play" aria-hidden="true"></span>
         </button>
+        <button id="record" type="button" class="btn btn-success">
```
Comment (xavierfav) [inline views/index.ejs:51]: As explained in #27, we want to export the loop created, not record the stream.
So I propose to put the button somewhere else, maybe at the top horizontal bar.

File: views/index.ejs [hunk 3]
```diff
@@ -132,6 +135,13 @@
         <input class="inputMessage" placeholder="Type here..." />
       </div>
     </div>
+    
+    <!--   EXPORT CONTAINER    -->
```
Comment (xavierfav) [inline views/index.ejs:139]: The idea was to make the user download the file directly.
Here it is not clear how to get the audio clip.

**Review by xavierfav — CHANGES_REQUESTED** [review]
The recording works, nice!
However this is not the functionality proposed in #27.
An export button should start the recording automatically and then make the user download the file.

File: views/index.ejs (hunk 2 above)
Comment (DerouineauNicolas) [reply views/index.ejs:51]: I'm personally in favor of the record button next to the play button. It allows to export more than one loop, and the bpm is known so exporting to a DAW is easy.

File: static/javascripts/app.js (hunk 1 above)
Comment (DerouineauNicolas) [reply static/javascripts/app.js]: Agreed, will be fixed.

File: views/index.ejs (hunk 3 above)
Comment (DerouineauNicolas) [reply views/index.ejs:139]: I added a download link next to the audio players in my latest commits.

File: static/javascripts/app.js [hunk 4]
```diff
@@ -131,12 +136,45 @@ function CheckAndTrigerPlayPause() {
 //  }
 //})
 
+function CheckAndTrigerRecord() {
```
Comment (xavierfav) [inline static/javascripts/app.js]: Function should start with lower case.

File: static/javascripts/app.js [hunk 5]
```diff
@@ -131,12 +136,45 @@ function CheckAndTrigerPlayPause() {
...
   });
 }
 
+function RecordListener() {
+  $('#record').click(function () {
+    CheckAndTrigerRecord();
+  });
+}
+
+function OnDataAvailableInRecorderFunc(evt) {
```
Comment (xavierfav) [inline static/javascripts/app.js]: Function should start with lower case.

File: static/javascripts/app.js [hunk 6]
```diff
@@ -206,8 +244,15 @@ function init() {
   loadImpulseResponses();
 }
 
+
```
Comment (xavierfav) [inline static/javascripts/app.js]: Extra line

File: static/javascripts/app.js [hunk 7]
```diff
@@ -131,12 +136,45 @@ function CheckAndTrigerPlayPause() {
...
+  });
+}
+
+function OnDataAvailableInRecorderFunc(evt) {
+  // push each chunk (blobs) in an array
+  if(evt.data.size>0){
+    chunks.push(evt.data);    
+    var blob = new Blob(chunks, { 'type' : 'audio/ogg; codecs=opus' });
+    var soundSrc = URL.createObjectURL(blob);
+    var NewHtmlEl = '<audio src=' + soundSrc+' controls=controls></audio><a href=' + soundSrc + ' download="exported_loop.ogg">Download</a><br>';
```
Comment (xavierfav) [inline static/javascripts/app.js]: Variable should start with lower case.

**Review by xavierfav — CHANGES_REQUESTED** [review]
Very nice. Although, little change is requested to accept the PR.

A few inconsistencies in the coding style.

The way players are added is not very fancy. 
Maybe just adding a button for the download would be enough for now.

Also, I think it would be time to create namespaces to make the code more modular.
For instance, all the functionalities related to the recorder could be in a function object. All the methods and objects could be accessed from one object only.
We can leave that part for #43.

**Comment by DerouineauNicolas** [comment]
I've unified the audio downloader container css with the rest of the sequencer containers style.

Regarding your comment about the audio players, I think it is better to have them in the container alongside the download links. This allows the user to preview what he is about to download.

The only issue is that the standard audio html  tag can't be styled through regular CSS properties. So one solution could be to create our own audio controller (play/pause/ect..) alongside a download button. But this seems out of this PR scope.

File: static/javascripts/app.js [hunk 8]
```diff
@@ -277,6 +320,7 @@ function playNote(buffer, noteTime, startTime, endTime, gainNode) {
 
   voice.connect(gainNode)
   gainNode.connect(currentLastNode);
+  gainNode.connect(recordingDest);
```
Comment (xavierfav) [inline static/javascripts/app.js:323]: Is it ok to always connect the node here?
Or should we just connect if we are recording?

Comment (xavierfav) [reply static/javascripts/app.js:323]: I think it allows to get sounding samples that have started before hitting the record button, which is nice

**Review by xavierfav — APPROVED** [review]
Good job!
